aten/src/ATen/native/cuda/AmpKernels.cu

```diff
@@ -0,0 +1,105 @@
+                                     const Tensor& found_inf)
+{
+  TORCH_CHECK(scaled_grad.is_cuda(), "scaled_grad must be a CUDA tensor.");
+  TORCH_CHECK(inv_scale.is_cuda(), "inv_scale must be a CUDA tensor.");
+  TORCH_CHECK(found_inf.is_cuda(), "found_inf must be a CUDA tensor.");
+  TORCH_CHECK(inv_scale.numel() == 1, "inv_scale must be a 1-element tensor.");
+  TORCH_CHECK(found_inf.numel() == 1, "found_inf must be a 1-element tensor.");
+  TORCH_CHECK(inv_scale.scalar_type() == at::ScalarType::Float, "inv_scale must be a float tensor.");
+  TORCH_CHECK(found_inf.scalar_type() == at::ScalarType::Float, "found_inf must be a float tensor.");
+
+  // Act on scaled_grad in place.
+  auto iter = TensorIterator::unary_op(scaled_grad, scaled_grad);
+
+  AT_DISPATCH_FLOATING_TYPES_AND_HALF(
+    iter.dtype(),
+    "_amp_unscale_inf_check_kernel",
+    [&] {
+      auto* found_inf_ptr = found_inf.data_ptr<float>();
+      auto* inv_scale_ptr = inv_scale.data_ptr<float>();
+
+      gpu_kernel(iter, [found_inf_ptr, inv_scale_ptr] GPU_LAMBDA(scalar_t val) -> scalar_t {
+          auto fval = static_cast<float>(val);
+          if (!std::isfinite(fval)) {
+            *found_inf_ptr = 1.f;
+          }
+          const auto inv_scale = *inv_scale_ptr; // Every thread accesses inv_scale, but it will hit in cache.
+          return static_cast<scalar_t>(inv_scale == 1.f ? fval : fval*inv_scale);
+        });
+    });
+
+  return scaled_grad;
+}
+
+
+// amp_update_scale_kernel is launched with a single thread to compute the new scale.
+__global__ void amp_update_scale_kernel(float* current_scale,
+                                        float* found_inf,
+                                        float* new_scale,
+                                        double scale_growth_factor,
+                                        double scale_backoff_factor)
+{
+  *new_scale = (*found_inf) ? (*current_scale)*scale_backoff_factor : (*current_scale)*scale_growth_factor;
+}
+
+
+// amp_update_scale_kernel asynchronously updates the scale factor.
+//
+// Args:
+// current_scale:  A one-element torch.cuda.FloatTensor containing the current scale value.
+// found_inf:  A one-element torch.cuda.FloatTensor. If > 0, indicates that infs/nans were found by the relevant
+//             prior _amp_unscale_inf_check_cuda call, and 0 if no infs/nans were found.
+// scale_growth_factor:  Multiplier if no infs/NaNs were found (typically slightly > 1).
+// scale_backoff_factor:  Multiplier if infs/NaNs were found (typically 0.5).
+//
+// Returns:
+// new_scale:  A new one-element torch.cuda.FloatTensor containing the new recommended scale value.
+Tensor _amp_update_scale_cuda(const Tensor& current_scale,
+                              const Tensor& found_inf,
+                              double scale_growth_factor,
+                              double scale_backoff_factor)
+{
+  TORCH_CHECK(current_scale.is_cuda(), "current_scale must be a CUDA tensor.");
```

Comment:
this doesn't check that you have a `torch.cuda.FloatTensor`, this checks that `tensor.device.type` is CUDA (e.g. a sparse tensor passes).

You can check that the device type is cuda and the the layout is strided instead.

Reply:
Good call.  I'll have to double check the Python code of `AmpScaler.py` for places a sparse tensor might slip through the cracks.

Alternatively, I could just make it support sparse tensors, by acting on the tensor that stores the values of the sparse tensor (don't remember of the top of my head what it's called).  It's a simple pointwise op.  What do you think?  Do that now, or add that feature later if it's called for?

Reply:
After looking more closely at the code, I'm going to put a layout check on the scaled gradient, but push the scale layout check to the Python side whenever the user chooses to define a new scale.  As long as they don't, I'm the only person who's creating the scale and found_inf tensors (and those only ever have numel == 1 anyway) so I don't think I need exhaustive checks for all of them in the kernels.